### PR TITLE
sql: add pow<int, int> and div<int, int>

### PR DESCRIPTION
**Pull request by hainesc**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10538)

### CODE REVIEW

**Comment by hainesc** [comment]
#10405

**Comment by tamird** [comment]
It is not enough to write "Fixes #10405" as the entire commit message; please write a multi-line commit message:

"sql: add pow<int,int>

Fixes #10405."

Also, please add a test.

File: pkg/sql/parser/builtins.go [hunk 1]
```diff
@@ -1271,6 +1271,15 @@ var powImpls = []Builtin{
 		_, err := decimal.Pow(&dd.Dec, x, y, decimal.Precision)
 		return dd, err
 	}),
+	{
+		Types:      ArgTypes{TypeInt, TypeInt},
+		ReturnType: TypeInt,
+		fn: func(_ *EvalContext, args DTuple) (Datum, error) {
+			x := float64(*args[0].(*DInt))
+			y := float64(*args[1].(*DInt))
+			return NewDInt(DInt(math.Pow(x, y))), nil
```
Comment (knz) [inline pkg/sql/parser/builtins.go]: Also this is rather inefficient. If you know you're exponentiating an integer use exponentiation by squaring, see https://en.wikipedia.org/wiki/Exponentiation_by_squaring (use the iterative version, and simply return 0 if the exponent is negative)

**Comment by knz** [comment]
Also like tamir says:
1. change the commit title to the full text of the PR title
2. leave a (short) explanation of what motivates the change in the body of the commit message
3. modify the PR comment with "Fixes ..."
4. see my comment about the algorithm. Float pow() is too expensive.
4. also it's "does not", not "dose not"

**Comment by hainesc** [comment]
It seems builtin function `div` has the same problem:
```sql
> select div(9, 5);
pq: unknown signature for div: div(int, int)
```

Should I rebase here with a single commit or just open another issue?

I think there are too few tests for builtin functions.

**Comment by knz** [comment]
You can keep it in the same PR but then in a different commit. Also, do not forget to add tests.

**Comment by knz** [comment]
Reviewed 1 of 2 files at r3.
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 1290 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDOi7V0ZecX6Huyljl:-KWDOi7WpxRLTZRl6v_H:b-ljg8o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/parser/builtins.go#L1290)):*
> ```Go
> 		ReturnType: TypeInt,
> 		fn: func(_ *EvalContext, args DTuple) (Datum, error) {
> 			// Caculate the x^y by squaring. See:
> ```

"Calculate"

---

*[pkg/sql/parser/builtins.go, line 1294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDOuFOv0AjIX16S3Zx:-KWDOuFOv0AjIX16S3Zy:ba8hmuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/parser/builtins.go#L1294)):*
> ```Go
> 			x := int(*args[0].(*DInt))
> 			y := int(*args[1].(*DInt))
> 			if y < 0 {
> ```

1) Move the core of this logic into a new function `func intPow(x, y int64) int64` 
2) special case x == 0 (result 1 if y = 0, 0 otherwise)

---

*[pkg/sql/testdata/builtin_function, line 610 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDPQy86JSGTAdm0cpf:-KWDPQy9BIAHn8Fyrgiy:b-pb1l2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/testdata/builtin_function#L610)):*
> ```
> 
> query III
> SELECT pow(-2::int, 3::int), pow(2::int, 3::int), pow(2::int, -3::int)
> ```

Test zero values on either and both sides too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538)*

**Comment by hainesc** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 1290 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDOi7V0ZecX6Huyljl:-KWDsPSbiIUazPzaKi-p:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/parser/builtins.go#L1290)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> "Calculate"
</details>
Thanks!

---

*[pkg/sql/parser/builtins.go, line 1294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDOuFOv0AjIX16S3Zx:-KWDsFY-PiNwrHmSXyJi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/parser/builtins.go#L1294)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> 1) Move the core of this logic into a new function `func intPow(x, y int64) int64` 
> 2) special case x == 0 (result 1 if y = 0, 0 otherwise)
</details>
Done.

---

*[pkg/sql/testdata/builtin_function, line 610 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDPQy86JSGTAdm0cpf:-KWDsVTHYz9Jl93PEwFO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d59f0b966a906c2ba122dd0847ee7650ec848998/pkg/sql/testdata/builtin_function#L610)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Test zero values on either and both sides too.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/builtins.go, line 1679 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDtYhBLcSqynFIdcl_:-KWDtYhBLcSqynFIdcla:b-ampbfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3753734a5399e81dc49dc0d78b69848d5c7230eb/pkg/sql/parser/builtins.go#L1679)):*
> ```Go
> 
> func intPow(x, y int64) int64 {
> 	// Calculate the x^y by squaring. See:
> ```

remove "the"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538)*

**Comment by hainesc** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/builtins.go, line 1679 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-KWDtYhBLcSqynFIdcl_:-KWDwy1l8XTIBQq8M0VS:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/3753734a5399e81dc49dc0d78b69848d5c7230eb/pkg/sql/parser/builtins.go#L1679)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove "the"
</details>
Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-:-KWDxRLOLFsTakUvildJ:bnfp4nl)*

**Comment by knz** [comment]
Thanks for your contribution! :lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10538#-:-KWDyYD3bT36ITRyzwfF:b-r36vnl)*

**Comment by jseldess** [comment]
Description added to docs with https://github.com/cockroachdb/docs/pull/948.

**Pull request merged**